File: app/channels/order_chat_channel.rb
class OrderChatChannel < ApplicationCable::Channel
  def subscribed
    @order = Order.find(params[:order_id])

    unless allowed_user?
      reject
      return
    end

    stream_for @order
  end

  def receive(data)
    message = @order.messages.create!(
      content: data["message"],
      user: current_user
    )

    OrderChatChannel.broadcast_to(@order, {
      message: message.content,
      user: current_user.name
    })
  end

  private

  def allowed_user?
    current_user.id == @order.user_id ||
    current_user.id == @order.restaurant.owner_id
  end
end

File: app/javascript/channels/order_chat_channel.js
import consumer from "./consumer"

export const createOrderChat = (orderId) => {
  return consumer.subscriptions.create(
    { channel: "OrderChatChannel", order_id: orderId },
    {
      connected() {
        console.log("Connected to chat")
      },

      received(data) {
        console.log("Message:", data)
      },

      send_message(message) {
        this.perform("receive", { message: message })
      }
    }
  )
}
